### ISSUE TITLE
Microsoft Dynamics 365: Refresh integration action results in 500 error

### ISSUE THREAD

**Issue by jamesmacwhite** [verbb/formie#1170]
### Describe the bug

I'm looking at implementing the Microsoft Dynamics 365 integration to our Dynamics 365 CRM environment. I've followed the guide to [setup the integration connection](verbb.io/craft-plugins/formie/docs/v1/integrations/crm#microsoft-dynamics-365) and enabled it, when I use the refresh integration button it makes a request but then after a while it throws a 500 error. On production a max memory limit was hit, on development the request `/index.php?p=manage/actions/formie/integrations/form-settings` returned a 500 error.

Is there any detailed logging/debugging I can do to better understand the issue.

Happy to dig into it if you need a Dynamics CRM environment to help test things.

### Steps to reproduce

N/A

### Form settings

N/A

### Craft CMS version

3.7.59

### Plugin version

1.6.18

### Multi-site?

No

**Comment by engram-design** [comment]
The `/storage/logs/formie.log` file should contain details on the 500 error, but strange you're getting a timeout.

**Comment by jamesmacwhite** [comment]
I'll dig into the logs. Failing that I'll try and query the API requests in something like Postman to understand what's happening with responses. Are there any major requests I should be looking at first?

**Comment by engram-design** [comment]
If you fancy a bit of debugging, you can start here - https://github.com/verbb/formie/blob/9fbe2b82592fcd1b50e524332f6326926d07ed94/src/integrations/crm/MicrosoftDynamics365.php#L166-L169

But as per [`_getEntityFields`](https://github.com/verbb/formie/blob/9fbe2b82592fcd1b50e524332f6326926d07ed94/src/integrations/crm/MicrosoftDynamics365.php#L391) there's only really 2x API requests for each data object. There might be a _lot_ of fields being returned on your system though.

**Comment by jamesmacwhite** [comment]
Thank you. It could be a field size issue given the timeout.  I'll have to do some API requests to see the payload response.

**Comment by engram-design** [comment]
I'd try to comment out some of those objects, and maybe isolate just the `$contactFields` call. Then, if still timing out, in `_getEntityFields`, remove one of the API calls (`$this->request()` and the logic afterwards) to see if you can isolate where this does awry.

**Comment by jamesmacwhite** [comment]
Thanks! Useful steps to try, I'll feedback when I've done some debugging!

**Comment by jamesmacwhite** [comment]
So initial debugging suggests the request response is massive (well for a web request JSON response). The contacts requests was around 88 MB, which seems massively inefficient. Other entities are large but not as larger.

I ended up importing the JSON into Excel for analysis and it looks like these endpoints are returning all fields and the data for these entities (there was 5000 rows), that would be why the data is so large. I don't know if there should be pagination or limits in place by default, but this is my first step into Dynamics 365 API related areas.

It would seems in our environment the endpoints like `{crmUrl}/api/data/v9.0/contacts` are returning all fields under the entity but with populated records. I don't know if that's the desired behaviour by default, or if limits and pagination should be implemented.

**Comment by engram-design** [comment]
Not being a Dynamics expert myself, that does seem interesting. How many custom fields do you have on contacts

I imagine it's the following call https://github.com/verbb/formie/blob/edfa833e914c1793cfa670957b33ce8603cf5eaa/src/integrations/crm/MicrosoftDynamics365.php#L396

```
/EntityDefinitions(LogicalName='contact')?$select=Attributes&$expand=Attributes($select=AttributeType,IsCustomAttribute,IsValidForCreate,IsValidForUpdate,CanBeSecuredForCreate,CanBeSecuredForUpdate,LogicalName,SchemaName,DisplayName,RequiredLevel)
```

Where we do indeed show all the attributes for a contact, but my understanding of that is it's just the definitions of them. That's not getting the data (which would be a call to `/contacts`)

**Comment by jamesmacwhite** [comment]
Trying this endpoint manually: `{crmUrl}/api/data/v9.0/EntityDefinitions(LogicalName='contact')?$select=Attributes&$expand=Attributes($select=AttributeType,IsCustomAttribute,IsValidForCreate,IsValidForUpdate,CanBeSecuredForCreate,CanBeSecuredForUpdate,LogicalName,SchemaName,DisplayName,RequiredLevel)` loads within a second, so that seems fine in this case.

**Comment by engram-design** [comment]
Ah, I think I know what's happening. It would be this: https://github.com/verbb/formie/blob/edfa833e914c1793cfa670957b33ce8603cf5eaa/src/integrations/crm/MicrosoftDynamics365.php#L574

Where that's used to fetch all contacts (and other objects) for you to be able to pick from for Picklist-type fields (picking a dynamics contact for the "Owner" field for example). That's your [5000 limit too](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/query-data-web-api#limits-on-number-of-table-rows-entities-returned). 

Fixed for the next release. To get this early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-3 as 1.6.18",
  "...": "..."
}
```

Then run `composer update`.

**Comment by jamesmacwhite** [comment]
Ah thanks! That would certainly explain the timeout if it trying to fetch all records, as it takes several minutes, running manually anywhere. I don't think I'll see a 88 MB JSON request again in debugging for a while!

**Comment by engram-design** [comment]
Fixed in [2.0.18](https://github.com/verbb/formie/releases/tag/2.0.18) and [1.6.19](https://github.com/verbb/formie/releases/tag/1.6.19)

**Comment by jamesmacwhite** [comment]
Confirmed. This now allows the refresh integration action to complete, so it was the large API request.
